File: R/example_usage.R
# =============================================================================
# Specific Affinity: R Examples
# =============================================================================
# This script demonstrates how to use the Specific Affinity framework in R
# for entity resolution and table matching.
# =============================================================================

# Load required packages
library(dplyr)
library(data.table)

# Source the main module
source("specific_affinity.R")

# =============================================================================
# EXAMPLE 1: Match Two Tables (Vendor Matching)
# =============================================================================

example_match_tables <- function() {
  cat("\n", strrep("=", 70), "\n")
  cat("EXAMPLE 1: Matching Two Tables\n")
  cat(strrep("=", 70), "\n\n")

  # Create sample reference data (master vendor list)
  master_vendors <- data.frame(
    vendor_id = c("V001", "V002", "V003", "V004", "V005"),
    vendor_name = c(
      "Netflix Inc",
      "Spotify Technology",
      "Amazon Web Services",
      "Microsoft Corporation",
      "Google Cloud Platform"
    ),
    stringsAsFactors = FALSE
  )

  # Create sample data to match (incoming transactions)
  incoming_data <- data.frame(
    record_id = c("R001", "R002", "R003", "R004", "R005", "R006"),
    merchant_name = c(
      "NETFLIX STREAMING",
      "Spotify Premium",
      "AWS Amazon",
      "MSFT Microsoft",
      "Unknown Vendor XYZ",
      "GOOGLE CLOUD"
    ),
    stringsAsFactors = FALSE
  )

  cat("Master Vendors:\n")
  print(master_vendors)
  cat("\nIncoming Data to Match:\n")
  print(incoming_data)

  # Run matching
  results <- match_tables(
    table_a = master_vendors,
    table_b = incoming_data,
    id_col_a = "vendor_id",
    id_col_b = "record_id",
    text_col_a = "vendor_name",
    text_col_b = "merchant_name",
    similarity_threshold = 0.4
  )

  # View results
  cat("\nMatch Results:\n")
  print(results$results[, .(table_b_id, merchant_name, matched_table_a_id,
                            matched_text, similarity_score, match_status)])

  return(results)
}

# =============================================================================
# EXAMPLE 2: Cluster Similar Records (Prime Table)
# =============================================================================

example_create_clusters <- function() {
  cat("\n", strrep("=", 70), "\n")
  cat("EXAMPLE 2: Clustering Similar Records\n")
  cat(strrep("=", 70), "\n\n")

  # Create sample transaction data with vendor name variations
  transactions <- data.frame(
    transaction_id = paste0("T", sprintf("%03d", 1:20)),
    memo = c(
      # Netflix variations
      "NETFLIX.COM", "NETFLIX SUBSCRIPTION", "NETFLIX INC", "NETFLIX STREAMING",
      # Spotify variations
      "SPOTIFY USA", "SPOTIFY PREMIUM", "SPOTIFY.COM",
      # Amazon variations
      "AMAZON.COM", "AMZN MKTP US", "AMAZON PRIME", "AMZN DIGITAL",
      # Uber variations
      "UBER TRIP", "UBER *EATS", "UBER TECHNOLOGIES",
      # Starbucks variations
      "STARBUCKS STORE 1234", "STARBUCKS COFFEE", "STARBUCKS #5678",
      # One-off transactions
      "LOCAL PIZZA SHOP", "JOE'S AUTO GARAGE", "RANDOM STORE"
    ),
    amount = c(15.99, 15.99, 15.99, 15.99,
               9.99, 9.99, 9.99,
               47.50, 23.99, 14.99, 4.99,
               24.50, 32.00, 18.75,
               5.75, 6.25, 4.50,
               25.00, 450.00, 15.00),
    stringsAsFactors = FALSE
  )

  cat("Sample Transactions:\n")
  print(transactions)

  # Create prime table (cluster similar records)
  result <- create_prime_table(
    df = transactions,
    id_col = "transaction_id",
    text_col = "memo",
    similarity_threshold = 0.4
  )

  # View clusters
  cat("\nClustered Results:\n")
  clustered <- result$prime_table[!is.na(cluster_id), .(transaction_id, memo, cluster_id)]
  clustered <- clustered[order(cluster_id)]
  print(clustered)

  # Show cluster summary
  cat("\nCluster Summary:\n")
  cluster_summary <- result$prime_table[!is.na(cluster_id),
                                         .(records = .N,
                                           sample_memo = first(memo)),
                                         by = cluster_id][order(-records)]
  print(cluster_summary)

  return(result)
}

# =============================================================================
# EXAMPLE 3: Match a Single Record
# =============================================================================

example_match_single <- function() {
  cat("\n", strrep("=", 70), "\n")
  cat("EXAMPLE 3: Match a Single Record\n")
  cat(strrep("=", 70), "\n\n")

  # Create reference data
  reference <- data.frame(
    id = c("V001", "V002", "V003", "V004", "V005"),
    name = c(
      "Netflix Inc",
      "Spotify Technology",
      "Amazon Web Services",
      "Microsoft Corporation",
      "Starbucks Coffee Company"
    ),
    stringsAsFactors = FALSE
  )

  # Create blocking keys and weights from reference
  blocking_keys <- create_blocking_keys(reference, "id", "name")
  weights <- calculate_weights(blocking_keys)

  # Test matching single records
  test_texts <- c(
    "NETFLIX PAYMENT",
    "Spotify Music Premium",
    "AMZN AWS",
    "Random Unknown Company"
  )

  for (text in test_texts) {
    cat(sprintf("\nQuery: '%s'\n", text))

    result <- match_single(
      text = text,
      blocking_keys = blocking_keys,
      weights = weights,
      reference_df = reference,
      id_col = "id",
      text_col = "name",
      similarity_threshold = 0.3,
      top_n = 3
    )

    if (!is.na(result$matched_id[1])) {
      print(result[, .(matched_id, matched_text, similarity_score, match_status)])
    } else {
      cat("  No matches found\n")
    }
  }
}

# =============================================================================
# EXAMPLE 4: Using dplyr Workflow
# =============================================================================

example_dplyr_workflow <- function() {
  cat("\n", strrep("=", 70), "\n")
  cat("EXAMPLE 4: dplyr Integration\n")
  cat(strrep("=", 70), "\n\n")

  # Create sample data
  vendors <- tibble(
    vendor_id = paste0("V", 1:5),
    vendor_name = c("Apple Inc", "Google LLC", "Microsoft Corp",
                    "Amazon.com Inc", "Facebook Meta")
  )

  transactions <- tibble(
    txn_id = paste0("T", 1:8),
    merchant = c("APPLE STORE", "GOOGLE PLAY", "MSFT OFFICE",
                 "AMZN SHOPPING", "META PLATFORMS", "APPLE MUSIC",
                 "UNKNOWN VENDOR", "RANDOM SHOP")
  )

  # Match and join back with dplyr
  match_result <- match_tables(
    table_a = vendors,
    table_b = transactions,
    id_col_a = "vendor_id",
    id_col_b = "txn_id",
    text_col_a = "vendor_name",
    text_col_b = "merchant",
    similarity_threshold = 0.3
  )

  # Use dplyr to analyze results
  summary_by_status <- match_result$results %>%
    as_tibble() %>%
    group_by(match_status) %>%
    summarise(
      count = n(),
      avg_score = mean(similarity_score, na.rm = TRUE),
      .groups = "drop"
    )

  cat("Summary by Match Status:\n")
  print(summary_by_status)

  # Find matched transactions with vendor info
  matched_transactions <- match_result$results %>%
    as_tibble() %>%
    filter(match_status == "MATCHED") %>%
    select(txn_id, merchant, matched_table_a_id, matched_text, similarity_score)

  cat("\nMatched Transactions:\n")
  print(matched_transactions)

  return(match_result)
}

# =============================================================================
# EXAMPLE 5: Custom Stop Words
# =============================================================================

example_custom_stop_words <- function() {
  cat("\n", strrep("=", 70), "\n")
  cat("EXAMPLE 5: Custom Stop Words\n")
  cat(strrep("=", 70), "\n\n")

  # Data with common words that might cause false matches
  companies <- data.frame(
    id = c("C001", "C002", "C003"),
    name = c(
      "International Business Solutions Inc",
      "Global Services Group LLC",
      "Technology Solutions Company"
    ),
    stringsAsFactors = FALSE
  )

  incoming <- data.frame(
    id = c("I001", "I002"),
    name = c(
      "International Technology Corp",
      "Business Group Inc"
    ),
    stringsAsFactors = FALSE
  )

  # Custom stop words for business domain
  custom_stop_words <- c(
    DEFAULT_STOP_WORDS,
    "international", "global", "group", "solutions",
    "services", "technology", "business", "company"
  )

  cat("Without custom stop words:\n")
  result1 <- match_tables(
    companies, incoming,
    "id", "id", "name", "name",
    similarity_threshold = 0.3
  )
  print(result1$results[, .(id = table_b_id, name, matched_table_a_id, similarity_score)])

  cat("\nWith custom stop words:\n")
  result2 <- match_tables(
    companies, incoming,
    "id", "id", "name", "name",
    similarity_threshold = 0.3,
    stop_words = custom_stop_words
  )
  print(result2$results[, .(id = table_b_id, name, matched_table_a_id, similarity_score)])
}

# =============================================================================
# Run Examples
# =============================================================================

run_all_examples <- function() {
  example_match_tables()
  example_create_clusters()
  example_match_single()
  example_dplyr_workflow()
  example_custom_stop_words()

  cat("\n", strrep("=", 70), "\n")
  cat("ALL EXAMPLES COMPLETE\n")
  cat(strrep("=", 70), "\n")
}

# Uncomment to run:
# run_all_examples()

# Or run individual examples:
# example_match_tables()
# example_create_clusters()
# example_match_single()

File: examples/example_usage.py
"""
Example: Using Specific Affinity for Entity Resolution

This example demonstrates how to use the Specific Affinity framework
to cluster similar records and match new records against existing clusters.

The example uses synthetic transaction data to show vendor matching.
"""

import sys
from pathlib import Path

# Add parent directory to path for imports
sys.path.insert(0, str(Path(__file__).parent.parent))

import duckdb
from python.main import SpecificAffinity


def create_sample_data(con: duckdb.DuckDBPyConnection) -> None:
    """Create sample transaction data for demonstration."""
    print("Creating sample data...")

    # Create transactions table
    con.execute("""
    CREATE OR REPLACE TABLE transactions (
        transaction_id VARCHAR PRIMARY KEY,
        memo VARCHAR,
        amount DECIMAL(10,2),
        transaction_date DATE,
        customer_id VARCHAR
    )
    """)

    # Insert sample transactions with various vendor name variations
    sample_data = [
        # Netflix variations
        ("t001", "NETFLIX.COM", 15.99, "2024-01-15", "c001"),
        ("t002", "NETFLIX SUBSCRIPTION", 15.99, "2024-02-15", "c001"),
        ("t003", "NETFLIX.COM/ACCT", 15.99, "2024-03-15", "c001"),
        ("t004", "NETFLIX INC", 15.99, "2024-01-20", "c002"),
        ("t005", "NETFLIX STREAMING", 15.99, "2024-02-20", "c002"),

        # Spotify variations
        ("t006", "SPOTIFY USA", 9.99, "2024-01-10", "c001"),
        ("t007", "SPOTIFY PREMIUM", 9.99, "2024-02-10", "c001"),
        ("t008", "SPOTIFY.COM", 9.99, "2024-03-10", "c001"),
        ("t009", "SPOTIFY SUBSCRIPTION", 9.99, "2024-01-12", "c003"),

        # Amazon variations
        ("t010", "AMAZON.COM", 47.50, "2024-01-05", "c001"),
        ("t011", "AMZN MKTP US", 23.99, "2024-01-18", "c001"),
        ("t012", "AMAZON PRIME", 14.99, "2024-02-01", "c001"),
        ("t013", "AMAZON.COM AMZN", 89.99, "2024-02-15", "c002"),
        ("t014", "AMZN DIGITAL", 4.99, "2024-02-20", "c002"),

        # Uber variations
        ("t015", "UBER TRIP", 24.50, "2024-01-08", "c001"),
        ("t016", "UBER *EATS", 32.00, "2024-01-15", "c001"),
        ("t017", "UBER TECHNOLOGIES", 18.75, "2024-01-22", "c002"),
        ("t018", "UBER RIDE", 15.00, "2024-02-05", "c002"),

        # Starbucks variations
        ("t019", "STARBUCKS STORE 1234", 5.75, "2024-01-03", "c001"),
        ("t020", "STARBUCKS COFFEE", 6.25, "2024-01-10", "c001"),
        ("t021", "STARBUCKS #5678", 4.50, "2024-01-17", "c002"),
        ("t022", "STARBUCKS MOBILE", 7.00, "2024-01-24", "c003"),

        # Target variations
        ("t023", "TARGET STORE", 156.78, "2024-01-12", "c001"),
        ("t024", "TARGET.COM", 45.00, "2024-01-25", "c001"),
        ("t025", "TARGET #1234", 89.99, "2024-02-08", "c002"),

        # One-off transactions (should remain unclustered or form small clusters)
        ("t026", "LOCAL PIZZA SHOP", 25.00, "2024-01-20", "c001"),
        ("t027", "JOE'S GARAGE AUTO", 450.00, "2024-01-22", "c001"),
        ("t028", "CITY PARKING LOT", 15.00, "2024-01-25", "c002"),
        ("t029", "DOWNTOWN DELI", 12.50, "2024-02-01", "c003"),
        ("t030", "HARDWARE STORE INC", 67.89, "2024-02-05", "c001"),
    ]

    con.executemany("""
    INSERT INTO transactions (transaction_id, memo, amount, transaction_date, customer_id)
    VALUES (?, ?, ?, ?, ?)
    """, sample_data)

    print(f"Created {len(sample_data)} sample transactions.")


def create_new_records(con: duckdb.DuckDBPyConnection) -> None:
    """Create new records to match against the prime table."""
    print("Creating new records for inference...")

    con.execute("""
    CREATE OR REPLACE TABLE new_transactions (
        transaction_id VARCHAR PRIMARY KEY,
        memo VARCHAR,
        amount DECIMAL(10,2),
        transaction_date DATE,
        customer_id VARCHAR
    )
    """)

    new_data = [
        # Should match existing clusters
        ("n001", "NETFLIX BILLING", 15.99, "2024-04-15", "c004"),
        ("n002", "SPOTIFY MONTHLY", 9.99, "2024-04-10", "c004"),
        ("n003", "AMZN MARKETPLACE", 33.50, "2024-04-12", "c004"),
        ("n004", "UBER *TRIP", 22.00, "2024-04-08", "c004"),
        ("n005", "STARBUCKS APP", 8.50, "2024-04-05", "c004"),

        # May or may not match
        ("n006", "PRIME VIDEO", 8.99, "2024-04-01", "c004"),  # Might match Amazon
        ("n007", "SBUX REWARDS", 5.00, "2024-04-03", "c004"),  # Might match Starbucks

        # Should not match (new vendors)
        ("n008", "GROCERY MART", 125.00, "2024-04-07", "c004"),
        ("n009", "ELECTRIC COMPANY", 89.50, "2024-04-15", "c004"),
        ("n010", "WATER UTILITY", 45.00, "2024-04-15", "c004"),
    ]

    con.executemany("""
    INSERT INTO new_transactions (transaction_id, memo, amount, transaction_date, customer_id)
    VALUES (?, ?, ?, ?, ?)
    """, new_data)

    print(f"Created {len(new_data)} new transactions for matching.")


def run_basic_example():
    """Run a basic example of the Specific Affinity pipeline."""
    print("\n" + "=" * 70)
    print("EXAMPLE: Basic Entity Resolution Pipeline")
    print("=" * 70 + "\n")

    db_path = "/tmp/specific_affinity_example.duckdb"

    # Initialize the framework
    sa = SpecificAffinity(
        db_path=db_path,
        text_field="memo",
        id_field="transaction_id",
        source_table="transactions",
        similarity_threshold=0.4
    )

    with sa:
        # Create sample data
        create_sample_data(sa.con)
        create_new_records(sa.con)

        # Run the complete pipeline
        results = sa.run_pipeline(
            new_records_table="new_transactions",
            run_qa=True
        )

        # Show some cluster examples
        print("\n" + "=" * 50)
        print("CLUSTER EXAMPLES")
        print("=" * 50)

        clusters = sa.con.execute(f"""
        SELECT DISTINCT cluster_id
        FROM {sa.config.prime_table}
        WHERE cluster_id IS NOT NULL
        LIMIT 5
        """).fetchall()

        for (cluster_id,) in clusters:
            print(f"\nCluster: {cluster_id}")
            samples = sa.get_cluster_sample(cluster_id, limit=5)
            for record_id, memo in samples:
                print(f"  - [{record_id}] {memo}")

        # Test single record matching
        print("\n" + "=" * 50)
        print("SINGLE RECORD MATCHING")
        print("=" * 50)

        test_texts = [
            "NETFLIX PAYMENT",
            "SPOTIFY FAMILY PLAN",
            "RANDOM NEW VENDOR",
            "AMAZON SHOPPING"
        ]

        for text in test_texts:
            result = sa.match_text(text)
            print(f"\nQuery: '{text}'")
            if result.get("matched"):
                print(f"  Matched to cluster: {result['assigned_cluster_id']}")
                print(f"  Best match: '{result['matched_text']}'")
                print(f"  Score: {result['similarity_score']:.4f}")
            else:
                print(f"  No match found: {result.get('reason', 'Unknown')}")

    print("\n" + "=" * 70)
    print("EXAMPLE COMPLETE")
    print("=" * 70 + "\n")


def run_categorization_example():
    """Run an example with transaction categorization."""
    print("\n" + "=" * 70)
    print("EXAMPLE: Transaction Categorization")
    print("=" * 70 + "\n")

    db_path = "/tmp/specific_affinity_categorization.duckdb"

    # Initialize with categorization fields
    sa = SpecificAffinity(
        db_path=db_path,
        text_field="memo",
        id_field="transaction_id",
        source_table="transactions",
        similarity_threshold=0.4,
        amount_field="amount",
        date_field="transaction_date",
        group_field="customer_id"
    )

    with sa:
        # Create sample data
        create_sample_data(sa.con)

        # Run pipeline with categorization
        results = sa.run_pipeline(
            include_categorization=True,
            run_qa=True
        )

        # Show categorization results
        if "step_4" in results and not results["step_4"].get("skipped"):
            print("\n" + "=" * 50)
            print("CATEGORIZATION RESULTS")
            print("=" * 50)

            cat_results = sa.con.execute("""
            SELECT
                cluster_type,
                record_type,
                COUNT(*) as count
            FROM transactions_prime_categorized
            WHERE cluster_id IS NOT NULL
            GROUP BY cluster_type, record_type
            ORDER BY count DESC
            """).fetchall()

            for cluster_type, record_type, count in cat_results:
                print(f"  {cluster_type or 'N/A'} / {record_type or 'N/A'}: {count} records")

    print("\n" + "=" * 70)
    print("CATEGORIZATION EXAMPLE COMPLETE")
    print("=" * 70 + "\n")


def run_custom_stop_words_example():
    """Example showing how to customize stop words."""
    print("\n" + "=" * 70)
    print("EXAMPLE: Custom Stop Words")
    print("=" * 70 + "\n")

    db_path = "/tmp/specific_affinity_stopwords.duckdb"

    # Initialize with custom stop words
    custom_stops = {"store", "shop", "inc", "llc", "corp", "payment", "billing"}

    sa = SpecificAffinity(
        db_path=db_path,
        text_field="memo",
        id_field="transaction_id",
        source_table="transactions",
        similarity_threshold=0.4,
        stop_words=custom_stops
    )

    with sa:
        # Create sample data
        create_sample_data(sa.con)

        # Add more stop words dynamically
        sa.add_stop_words({"monthly", "subscription", "account"})

        # Run Step 1 only
        sa.create_prime_table()

        # Show token weights (custom stop words should be excluded)
        print("\nTop weighted tokens (custom stop words excluded):")
        tokens = sa.con.execute(f"""
        SELECT token, weight
        FROM {sa.config.weights_table}
        ORDER BY weight DESC
        LIMIT 10
        """).fetchall()

        for token, weight in tokens:
            print(f"  {token}: {weight:.4f}")

    print("\n" + "=" * 70)
    print("CUSTOM STOP WORDS EXAMPLE COMPLETE")
    print("=" * 70 + "\n")


if __name__ == "__main__":
    import argparse

    parser = argparse.ArgumentParser(description="Specific Affinity Examples")
    parser.add_argument(
        "--example",
        choices=["basic", "categorization", "stopwords", "all"],
        default="basic",
        help="Which example to run"
    )

    args = parser.parse_args()

    if args.example == "basic" or args.example == "all":
        run_basic_example()

    if args.example == "categorization" or args.example == "all":
        run_categorization_example()

    if args.example == "stopwords" or args.example == "all":
        run_custom_stop_words_example()
